### PR TITLE
Minter Access Control

### PR DESCRIPTION
**Pull request by NicolasMahe**
Hey guys,

This PR is the implementation of suggestions from the discussion of my business partner: https://github.com/rarible/protocol/discussions/11 and is the phase 2 of the proposal https://gov.rarible.org/t/proposal-to-optionally-restrict-the-mint-of-tokens/182.

This PR adds a new smart contract, `MinterAccessControl`, to restrict who can mint tokens.

The state variable `_minters` contains the addresses that are authorized to mint.
Functions `grantMinter` and `revokeMinter` add or remove an address to/from the authorized minter list. Only the smart contract's owner can execute these functions.

The state variable `minterAccessControlEnabled` controls the restriction of mint. `false` by default not to create breaking change.
Functions `enableMinterAccessControl` and `disableMinterAccessControl` enable and disable the restriction of mint. Only the smart contract's owner can execute these functions.

Finally, the function `isValidMinter` asserts if the minter is authorized only if the mint restriction is enabled. Return true for any minter if the mint restriction is disabled.
This function is called in the `mintAndTransfer` function of the contracts `ERC721LazyMinimal` and `ERC721Lazy` using the minter encoded in the `tokenId`.

The `MinterAccessControl` restriction is disabled by default the prevent breaking change. The contract owner has to call `enableMinterAccessControl` and add the minter's address with `grantMinter` function.

### CODE REVIEW

**Comment by artsyassets** [comment]
Thank you!

**Comment by basti4nos** [comment]
Nice one!

**Comment by eduardstal** [comment]
@evgenynacu Would this PR be in conflict with the fix we're deploying to restrict minting with private `createToken`?

**Comment by chusla** [comment]
> @evgenynacu Would this PR be in conflict with the fix we're deploying to restrict minting with private `createToken`?

came here to ask the same question!  any input would be much appreciated!  current private createtoken is a great start thanks for that.  This would give more flexibility on top of it ideally.

**Comment by evgenynacu** [comment]
Hey guys
Thanks for the PR
Team will give the feedback soon

Sorry for the delay

File: tokens/test/erc-721/Rarible.test.js [hunk 1]
```diff
@@ -322,6 +322,83 @@ contract("ERC721Rarible", accounts => {
...
+    const tokenId = minter + "b00000000000000000000001";
+    const tokenURI = "//uri";
+
+    await token.enableMinterAccessControl({from: tokenOwner});
+    assert.equal(await token.minterAccessControlEnabled(), true);
+
+    const signature = await getSignature(tokenId, tokenURI, creators([minter]), [], minter);
+
+    let whiteListProxy = accounts[5];
+    await token.setDefaultApproval(whiteListProxy, true, {from: tokenOwner});
```
Comment (disenotov) [inline tokens/test/erc-721/Rarible.test.js:360]: also in the last version of rarible tokens (where isPrivate variable appears) there is no setDefaultApproval() method, so you probably need to take that into consideration

File: tokens/contracts/access/MinterAccessControl.sol [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import "@openzeppelin/contracts-upgradeable/proxy/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/EnumerableSetUpgradeable.sol";
+
+contract MinterAccessControl is Initializable, OwnableUpgradeable {
+    using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+
+    EnumerableSetUpgradeable.AddressSet private _minters;
+    bool public minterAccessControlEnabled;
```
[the remark this answers is not in the document]
Comment (NicolasMahe) [reply tokens/contracts/access/MinterAccessControl.sol]: this has been changed in commit https://github.com/rarible/protocol-contracts/pull/105/commits/6a5e213b9ed8f38f84d2397dcf61771d038c4b8e

File: tokens/test/erc-1155/MinterAccessControl1155.test.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    await token.enableMinterAccessControl({from: tokenOwner});
+    assert.equal(await token.minterAccessControlEnabled(), true);
+
+    await expectThrow(
+      token.mintAndTransfer([tokenId, tokenURI, supply, creators([minter]), [], [zeroWord]], transferTo, mint, {from: minter})
+    );
+
+    // upgrade contract
+    const newInstance = await upgradeProxy(token.address, Testing);
```
[the remark this answers is not in the document]
Comment (NicolasMahe) [reply tokens/test/erc-1155/MinterAccessControl1155.test.js]: should be fixed in https://github.com/rarible/protocol-contracts/pull/105/commits/eb3607e22bf9f7bc4dae59c4b08fa945ea1f2ef7#diff-5e2ca5ca204052f7aee15e678eddb911242048bd4213ecf9fcd41c5bdda8cdd8

File: tokens/test/erc-721/Rarible.test.js (hunk 1 above)
Comment (NicolasMahe) [reply tokens/test/erc-721/Rarible.test.js]: removed in commit https://github.com/rarible/protocol-contracts/pull/105/commits/6a5e213b9ed8f38f84d2397dcf61771d038c4b8e

**Comment by NicolasMahe** [comment]
> Hello! sorry for the long wait, here are some comments feel free to ask any questions, gonna answer faster this time :)

Comments implemented 👍 

The state variable `minterAccessControlEnabled` has been removed in offer of `isPrivate`.

File: tokens/contracts/access/MinterAccessControl.sol [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        __MinterAccessControl_init_unchained();
+    }
+
+    function __MinterAccessControl_init_unchained() internal initializer {
+    }
+
+    /**
+     * @dev Add `_minter` to the list of allowed minters.
+     */
+    function grantMinter(address _minter) external onlyOwner {
```
Comment (SKonstantinS) [inline tokens/contracts/access/MinterAccessControl.sol:28]: I think is better to change "**addMinter**" for method name **grantMinter**

File: tokens/contracts/access/MinterAccessControl.sol [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        require(!_minters.contains(_minter), 'MinterAccessControl: Already minter');
+        _minters.add(_minter);
+        emit MinterGranted(_minter);
+    }
+
+    /**
+     * @dev Revoke `_minter` from the list of allowed minters.
+     */
+    function revokeMinter(address _minter) external onlyOwner {
```
Comment (SKonstantinS) [inline tokens/contracts/access/MinterAccessControl.sol:37]: I think is better to change "**removeMinter**" for method name **revokeMinter**

File: tokens/contracts/access/MinterAccessControl.sol [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        require(_minters.contains(_minter), 'MinterAccessControl: Not minter');
+        _minters.remove(_minter);
+        emit MinterRevoked(_minter);
+    }
+
+    /**
+     * @dev Returns `true` if `account` has been granted to minters.
+     */
+    function isValidMinter(address account) public view returns (bool) {
```
Comment (SKonstantinS) [inline tokens/contracts/access/MinterAccessControl.sol:46]: I think is better to change "isMinter" for method name **isValidMinter**

File: tokens/contracts/access/MinterAccessControl.sol [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import "@openzeppelin/contracts-upgradeable/proxy/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/EnumerableSetUpgradeable.sol";
+
+abstract contract MinterAccessControl is Initializable, OwnableUpgradeable {
+    using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+
+    EnumerableSetUpgradeable.AddressSet private _minters;
+    
+    event MinterGranted(address indexed account);
```
Comment (SKonstantinS) [inline tokens/contracts/access/MinterAccessControl.sol:14]: 1.I think is better to change "MinterGranted" for  MinterAdded
2.Lets think togrther, do we need to emit `operator` (_msgSender()), in event parameters like that:
event MinterAdded(address indexed operator, address indexed newMinter ) 
3.need test for cheking emit event

File: tokens/contracts/access/MinterAccessControl.sol [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    event MinterRevoked(address indexed account);
```
Comment (SKonstantinS) [inline tokens/contracts/access/MinterAccessControl.sol:15]: The same for **MinterRevoked** like for MinterGranted:
1. name MinterRemoved
2. emit operator
3.test

File: tokens/test/erc-1155/ERC1155RaribleUser.test.js [hunk 9]
```diff
@@ -177,6 +177,80 @@ contract("ERC1155RaribleUser", accounts => {
...
+    await expectThrow(
+      token.mintAndTransfer([tokenId, tokenURI, supply, creators([minter]), [], [zeroWord]], transferTo, mint, {from: minter})
+    );
+
+    await token.grantMinter(minter, {from: tokenOwner});
+    assert.equal(await token.isValidMinter(minter), true);
+    assert.equal(await token.isValidMinter(transferTo), false);
+
+    await token.mintAndTransfer([tokenId, tokenURI, supply, creators([minter]), [], [zeroWord]], transferTo, mint, {from: minter});
+		assert.equal(await token.uri(tokenId), "ipfs:/" + tokenURI);
```
Comment (SKonstantinS) [inline tokens/test/erc-1155/ERC1155RaribleUser.test.js:198]: Line 198-200. In This test we don`t need any additional checking. I think this code  interfere the main test idea check 'minter access control'
Please review another two tests with 'minter access control' and delete unnecessary code

**Comment by SKonstantinS** [comment]
Hi @NicolasMahe, I leave some comments, not critical, I hope You are agree. Together we`ll try to generate this code better)))

File: tokens/contracts/access/MinterAccessControl.sol (hunk 4 above)
Comment (NicolasMahe) [reply tokens/contracts/access/MinterAccessControl.sol]: done

File: tokens/contracts/access/MinterAccessControl.sol (hunk 5 above)
Comment (NicolasMahe) [reply tokens/contracts/access/MinterAccessControl.sol]: done and done

File: tokens/contracts/access/MinterAccessControl.sol (hunk 6 above)
Comment (NicolasMahe) [reply tokens/contracts/access/MinterAccessControl.sol]: done

File: tokens/contracts/access/MinterAccessControl.sol (hunk 7 above)
Comment (NicolasMahe) [reply tokens/contracts/access/MinterAccessControl.sol]: 3 points done

File: tokens/contracts/access/MinterAccessControl.sol (hunk 8 above)
Comment (NicolasMahe) [reply tokens/contracts/access/MinterAccessControl.sol]: done

File: tokens/test/erc-1155/ERC1155RaribleUser.test.js (hunk 9 above)
Comment (NicolasMahe) [reply tokens/test/erc-1155/ERC1155RaribleUser.test.js]: the `assert.equal(await token.uri(tokenId), "ipfs:/" + tokenURI);` has been removed.

do you also want to remove the following:
```
assert.equal(await token.balanceOf(transferTo, tokenId), mint);
assert.equal(await token.balanceOf(minter, tokenId), 0);
```

They should not be needed if `mintAndTransfer` works as intended but they are here to make sure the minter access control didn't break something.

**Comment by NicolasMahe** [comment]
> Hi @NicolasMahe, I leave some comments, not critical, I hope You are agree. Together we`ll try to generate this code better)))

Modification done 👍  One conversation open for more modification.

File: tokens/test/erc-1155/ERC1155RaribleUser.test.js (hunk 9 above)
Comment (SKonstantinS) [reply tokens/test/erc-1155/ERC1155RaribleUser.test.js]: Hi @NicolasMahe, thanks for good work!
Ok agree let's leave this 2 lines in test:
`assert.equal(await token.balanceOf(transferTo, tokenId), mint);
assert.equal(await token.balanceOf(minter, tokenId), 0);`

File: tokens/test/contracts/access/MinterAccessControlTestV2.sol [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import "../../../contracts/access/MinterAccessControl.sol";
+
+contract MinterAccessControlTestV2 is MinterAccessControl {
+
+    function initialize() external initializer {
+        __Ownable_init_unchained();
+        __MinterAccessControl_init_unchained();
+    }
+
+    function version() public view returns (uint256) {
```
Comment (SKonstantinS) [inline tokens/test/contracts/access/MinterAccessControlTestV2.sol:14]: For version identification we use smth like that:
bytes4 constant public V2 = bytes4(keccak256("V2"));
You can see it: exchange-v2/contracts/LibOrderDataV2.sol
So may You use this constant in test
I think there is no error here, but probably we should use the same approach

Comment (NicolasMahe) [reply tokens/test/contracts/access/MinterAccessControlTestV2.sol]: updated: https://github.com/rarible/protocol-contracts/pull/105/commits/af18208855fd179838ea47e014d696363c0df14e

Comment (SKonstantinS) [reply tokens/test/contracts/access/MinterAccessControlTestV2.sol]: Thank You for your work, I`ll review

Comment (NicolasMahe) [reply tokens/test/contracts/access/MinterAccessControlTestV2.sol]: @SKonstantinS waiting for your review ;)

**Comment by disenotov** [comment]
merged into https://github.com/rarible/protocol-contracts/tree/RPC-132-minter-role branch for now, gonna merge into master soon, thanks!

**Comment by evgenynacu** [comment]
@NicolasMahe merged into master too

**Comment by bobwith2bees** [comment]
@disenotov Question on Minter Access control: can you enable this _after_ a contract has been deployed with isPrivate=false?

It appears it is enabled by the isPrivate boolean, but I don't see an "enable" function to change the setting via any of the ABI's.  What is the name of the function to enable it?

I can access the MinterAccessControl functions (isMinter, addMinter etc).

**Comment by bobwith2bees** [comment]
@disenotov  That's what I needed to know.  Thanks for the clarification.

**Pull request merged**